### ISSUE TITLE
Add a Padding layer to Caffe

### ISSUE THREAD

**Issue by twerdster** [BVLC/caffe#6294]
We have developed a padding layer for Caffe similar in functionality to Keras, CoreML and others. The implementation is currently for GPU (Cuda) and CPU and has been working without issue for around a year. Is this likely to be accepted as a PR? If it is then we can make the effort to add the necessary layer tests.

This is what the proto message looks like.

`message PaddingParameter {
  enum PadMethod {
    CONSTANT = 0; // constant border fill, can choose what value
    REPEAT = 1;   // repeat the values along the border (i.e. CLAMP)
    CLAMP = 2;    // same as REPEAT
    REFLECT = 3;  // mirror about the edge
    WRAP = 4;     // wrap the values around from the other side
  }
  optional PadMethod pad_method = 1 [default = REPEAT]; // The padding method
  // Explicit padding for every side is given
  // For reference: top left corner is the origin
  optional uint32 pad_l = 2 [default = 0]; // number of columns to add to the left
  optional uint32 pad_r = 3 [default = 0]; // number of columns to add to the right
  optional uint32 pad_t = 4 [default = 0]; // number of rows to add to the top 
  optional uint32 pad_b = 5 [default = 0]; // number of rows to add to the bottom
  optional float pad_value = 6 [default = 0]; // if pad_method is CONSTANT then use this value

  enum Engine {
    DEFAULT = 0;
    CAFFE = 1;
  }
  optional Engine engine = 7 [default = DEFAULT];
}`

**Comment by twmht** [comment]
@twerdster 

when do we need padding layer?

**Comment by twerdster** [comment]
Style transfer filters such as [fast-neural-style](https://github.com/jcjohnson/fast-neural-style/blob/master/fast_neural_style/models.lua)  can make use of it.

Also, the use of symmetric zero padding as the only option available seems to be an unnecessary limitation. Many image processing algorithms require non-zero boundary conditions. 

Here are paddings provided in some other frameworks:
[Torch](https://github.com/torch/nn/blob/master/doc/convolution.md#nn.SpatialZeroPadding)
[Tensorflow](https://www.tensorflow.org/api_docs/python/tf/pad) 
[Keras](https://keras.io/layers/convolutional/#conv2d)
[CoreML](https://apple.github.io/coremltools/coremlspecification/sections/NeuralNetwork.html#validpadding)
[PyTorch](http://pytorch.org/docs/master/_modules/torch/nn/modules/padding.html)

**Comment by twmht** [comment]
@twerdster 

Can you submit a PR for that?

**Comment by twerdster** [comment]
@twmht 
I would like to first know whether this is something that would be accepted. It requires non trivial effort on our part (which we are happy to do if it has a chance to be accepted)  to create the test cases and gradient check tests. Also, it requires adding a new caffe layer ID which has not happened for a considerably period of time so I presume this is simply uncommon.

Could you give me some feedback and what you see the chances of this happening?

**Comment by twmht** [comment]
@twerdster 

I think this has a great chance to be accepted, since it has been used in many of frameworks.

So, you have not completed the code?

**Comment by twerdster** [comment]
@twmht Are you a Caffe maintainer? The padding code works in our networks without issue. The only thing left to build are the layer tests and gradient checks as required for a layer contribution.

**Comment by twmht** [comment]
@twerdster 

I am not a caffe maintainer. I am just interested in your work.

**Comment by Noiredd** [comment]
@twerdster Sorry for making you wait so long with the answer.

Yes, I think such layer could be useful. In general we want actual processing layers to handle padding themselves (as a part of the algorithm), since an explicit padding layer in most cases would only use up that precious RAM unnecessarily. However, I think explicit padding layer could be useful - for example in networks that output images (autoencoders, segmentation etc.).

But in general, if we have means to cut from a blob (crop layer), it would be good to be able to add to it too. So if you decide to dedicate your time and effort to implementing the tests, I will gladly review and merge such PR. I suggest you submit it as is for now, even without the tests - so we can take a look at the layer itself while you code the tests.

**Comment by ujsyehao** [comment]
@Noiredd Hi, Dose caffe still support update? 
I implement **ResizeBilinear layer in caffe(equal to tf.image.resize_bilinear operation)** and I have tested the added layer, Can it to be accepted as a PR? 
Additionally, caffe only supports **symmetric padding**, it is different from tensorflow, If I want to modify convolution and pooling padding code **to make it support for asymmetric padding**, 
Which part should I modify?

**Comment by ujsyehao** [comment]
@Noiredd Hi, Can you give me some advice?

**Comment by taojiang82** [comment]
@ujsyehao Hi, can you share the ResizeBilinear layer code?

**Comment by Noiredd** [comment]
@ujsyehao If you've got the code working - post a PR and let the community see it. I think such a layer might be interesting to some.

Regarding the asymmetric padding, I don't think this will be easy, especially in the N-D case. @twerdster's solution would come in immensely handy in your situation.

**Comment by 17cspence** [comment]
I just (perhaps stupidly) submitted a pull request for a pad layer, #6506. *Now* I see this issue. But I've had the code for a while. It only handles the 2D case. As @Noiredd commented, N-D is probably harder.

**Comment by ujsyehao** [comment]
@taojiang82 I have submitted a PR and you can see the code.

**Comment by ujsyehao** [comment]
@Noiredd Hi, I have posted a PR, please refer to #6508, Thank you for your reply!

**Comment by snownus** [comment]
@twerdster , have added the padding layer???

**Comment by twerdster** [comment]
Unfortunately I wont be adding the layer. Sorry.

**Comment by 17cspence** [comment]
@snownus, as mentioned above, I've submitted PR #6506 for a pad layer. Compared to @twerdster's, my code only allows padding that's the same on all sides. And for modes, twerdster's modes and mine are:
twerdster                    mine
CONSTANT               ZERO, but that's the only value my code allows, twerdster allows any value
REPEAT or CLAMP   REPLICATE
REFLECT                  REFLECT and REFLECT_101, don't know which version twerdster implemented
WRAP                       Mine doesn't have this.

For a pad width of 2, my REFLECT does abcdefg... -> baabcdefg ..., while REFLECT_101 does abcdefg... -> cbabcdefg ... The names REPLICATE, REFLECT, and REFLECT_101 are taken from the OpenCV library.

**Comment by snownus** [comment]
@17cspence ， you mean padding_l == padding_r == padding_t == padding_b?

**Comment by 17cspence** [comment]
@snownus, yes, that's what I mean.

**Comment by snownus** [comment]
@17cspence ， ok, thanks.

**Comment by siyiding1216** [comment]
> @snownus, as mentioned above, I've submitted PR #6506 for a pad layer. Compared to @twerdster's, my code only allows padding that's the same on all sides. And for modes, twerdster's modes and mine are:
> twerdster -- mine
> CONSTANT -- ZERO, but that's the only value my code allows, twerdster allows any value
> REPEAT or CLAMP -- REPLICATE
> REFLECT -- REFLECT and REFLECT_101, don't know which version twerdster implemented
> WRAP -- Mine doesn't have this.
> 
> For a pad width of 2, my REFLECT does abcdefg... -> baabcdefg ..., while REFLECT_101 does abcdefg... -> cbabcdefg ... The names REPLICATE, REFLECT, and REFLECT_101 are taken from the OpenCV library.

Where did you see twerdster`s code?

**Comment by 17cspence** [comment]
@siyiding1216, I did not see twerdster's code. Enough time has gone by that I'm not quite sure what I was thinking. I suspect I just worded that comment imprecisely. You probably saw twerdster's comment that they won't submit this code:

> 
> Unfortunately I wont be adding the layer. Sorry.

**Comment by 17cspence** [comment]
@siyiding1216, to be complete I should mention that my list of twerdster's modes was from the first March 15 post, describing the proto message.

**Comment by vikramambrose** [comment]
@twerdster or anyone else who maintains this repo.

To be clear, was it the 6506 PR padding layer specifically that was denied or are you not interested adding a padding layer at all even if someone were to write a proper padding layer respecting the new blob vector<int> shape api?

Or would you want to see this functionality added to the conv layer instead?

**Comment by zcl912** [comment]
> We have developed a padding layer for Caffe similar in functionality to Keras, CoreML and others. The implementation is currently for GPU (Cuda) and CPU and has been working without issue for around a year. Is this likely to be accepted as a PR? If it is then we can make the effort to add the necessary layer tests.
> 
> This is what the proto message looks like.
> 
> ```
> message PaddingParameter {
>   enum PadMethod {
>     CONSTANT = 0; // constant border fill, can choose what value
>     REPEAT = 1;   // repeat the values along the border (i.e. CLAMP)
>     CLAMP = 2;    // same as REPEAT
>     REFLECT = 3;  // mirror about the edge
>     WRAP = 4;     // wrap the values around from the other side
>   }
>   optional PadMethod pad_method = 1 [default = REPEAT]; // The padding method
>   // Explicit padding for every side is given
>   // For reference: top left corner is the origin
>   optional uint32 pad_l = 2 [default = 0]; // number of columns to add to the left
>   optional uint32 pad_r = 3 [default = 0]; // number of columns to add to the right
>   optional uint32 pad_t = 4 [default = 0]; // number of rows to add to the top 
>   optional uint32 pad_b = 5 [default = 0]; // number of rows to add to the bottom
>   optional float pad_value = 6 [default = 0]; // if pad_method is CONSTANT then use this value
> 
>   enum Engine {
>     DEFAULT = 0;
>     CAFFE = 1;
>   }
>   optional Engine engine = 7 [default = DEFAULT];
> }
> ```


prototxt like ?:
layer {
  name: "Pad273"
  type: "Pad"
  bottom: "mul_blob150"
  top: "Pad_blob273"
  pad_param {
    type: ZERO
    pad_l: 1
    pad_r: 1
    pad_t: 1
    pad_b: 1
  }
}
